README.md
# VScript Library
[![ver][]](CHANGELOG.txt)

[ver]: https://img.shields.io/badge/vs__library-v2.43.29-informational

High-performance vscript libraries; written mainly for CS:GO, compatible with Portal 2.

## Documentation
See [Documentation.md](Documentation.md)

## Installation
Download the library file you will be using, and place it in your vscripts directory `/csgo/scripts/vscripts/` - the source files are not required
- [`vs_math.nut`][vs_math]: Standalone math library. Game independent.
- [`vs_events.nut`][vs_events]: Standalone lightweight game events library. CSGO only.
- [`vs_library.nut`][vs_library]: All libraries. Includes unique utility functions in addition to the math and events libraries.
- [`glow.nut`][glow]: Standalone glow object manager.

[vs_math]: https://raw.githubusercontent.com/samisalreadytaken/vs_library/master/vs_math.nut
[vs_events]: https://raw.githubusercontent.com/samisalreadytaken/vs_library/master/vs_events.nut
[vs_library]: https://raw.githubusercontent.com/samisalreadytaken/vs_library/master/vs_library.nut
[glow]: https://raw.githubusercontent.com/samisalreadytaken/vs_library/master/glow.nut

## Usage
Include the particular library file at the beginning of your script: `IncludeScript("vs_library")`

Done!

It only needs to be included once in the lifetime of the map running in the server. Including it more than once does not affect anything.

### Math library
The math library is based on Source engine. Working with this makes moving from and to Source development very easy.

### EventQueue
Utility for asynchronous function execution.
```cs
function fn2() { print("2\n"); }
function fn3() { print("3\n"); }

print("1\n");
VS.EventQueue.AddEvent( fn2, 1.0, this );
VS.EventQueue.AddEvent( fn3, 2.0, this );
VS.EventQueue.CancelEventsByInput( fn2 );
```

### Extended player
Use `ToExtendedPlayer()` to access some of the missing player functions in CSGO such as `EyeAngles()`, `GetUserID()`, `GetPlayerName()` and `GetFOV()`/`SetFOV()`. See the [documentation](/Documentation.md#f_ToExtendedPlayer) for details.

```cs
local player = ToExtendedPlayer( VS.GetPlayerByIndex(1) );

print(format( "Draw view frustum of [%s] %s\n", player.GetNetworkIDString(), player.GetPlayerName() ));

local aspectRatio = 16.0/9.0;

VS.DrawViewFrustum( player.EyePosition(), player.EyeForward(), player.EyeRight(), player.EyeUp(),
	VS.CalcFovX( player.GetFOV(), aspectRatio * (3.0/4.0) ), aspectRatio,
	7.0, 16.0, 255, 0, 0, false, 5.0 );

DebugDrawBoxAngles( player.EyePosition(), Vector(2,-0.5,-0.5), Vector(16,0.5,0.5), player.EyeAngles(), 0, 255, 0, 16, 5.0 );
```

### Automatic player info acquisition
[![](https://img.shields.io/badge/video-red?logo=youtube)](https://www.youtube.com/watch?v=JGnBQ1lwzzg)

The game events library completely automates player userid, SteamID and Steam name acquisition; and also exposes game event listener registration from script with `VS.ListenToGameEvent()`.

This requires setting up these 2 entities:

```
logic_eventlistener:
	targetname: vs.eventlistener

point_template:
	Entity Scripts: vs_eventlistener.nut
	Template01: vs.eventlistener
```

where `vs_eventlistener.nut` file contents should read:

```cpp
IncludeScript("vs_events");
VS.Events.InitTemplate(this);
```

Player info will automatically be put in their script scope.

```cs
local userid = 2;
local player = VS.GetPlayerByUserid( userid );
local scope = player.GetScriptScope();

printl( scope.userid );
printl( scope.networkid );
printl( scope.name );
```

This also enables `VS.ListenToGameEvent()` and `VS.StopListeningToAllGameEvents()` functions to manage event listeners dynamically from script.

```cs
IncludeScript("vs_events");

VS.ListenToGameEvent( "bullet_impact", function( event )
{
	local position = Vector( event.x, event.y, event.z );
	local player = VS.GetPlayerByUserid( event.userid );

	DebugDrawLine( player.EyePosition(), position, 255, 0, 0, false, 2.0 );
	DebugDrawBox( position, Vector(-2,-2,-2), Vector(2,2,2), 255, 0, 255, 127, 2.0 );
}, "DrawImpact" );

VS.ListenToGameEvent( "player_disconnect", function( event )
{
	local player = VS.GetPlayerByUserid( event.userid );
	local origin = player.GetOrigin();

	print(format( "Player disconnected at %f %f %f\n", origin.x, origin.y, origin.z ));
	DebugDrawBox( origin, Vector(-16,-16,0), Vector(16,16,72), 255, 0, 0, 127, 10.0 );

}, "ShowDisconnectLocation", 1 );
```

#### Use on dedicated servers
It is not possible to get the Steam name and SteamIDs of human players that were connected to a server prior to a map change because the player_connect event is fired only once when a player connects to the server. This data will only be available for players that connect to the server while your map is running.

This issue is fixed in listen servers, and players are guaranteed to have valid SteamIDs.

---

### Why is it minified?
Minification lets me micro optimise the library further by reusing local variables and strings, getting rid of line info, reducing stack size and memory usage. Applying such changes to the source would make it unreadable.

The code inside the minified files are otherwise identical to those inside the `src` directory.

### Should I minify my scripts?
No, this is completely unnecessary and give no noticable benefit whatsoever.

## Changelog
See [CHANGELOG.txt](CHANGELOG.txt)

## Licence
You are free to use, modify and share this library under the terms of the MIT License. The only condition is keeping the copyright notice, and stating whether or not the code was modified. See [LICENSE](LICENSE) for details.

[![](http://hits.dwyl.com/samisalreadytaken/vs_library.svg)](https://hits.dwyl.com/samisalreadytaken/vs_library)

________________________________

## See also
* [**Notepad++ syntax highlighter**][npp]

[npp]: https://gist.github.com/samisalreadytaken/5bcf322332074f31545ccb6651b88f2d


CHANGELOG.txt

	github.com/samisalreadytaken/vs_library

--------------------------------

2023-02-02 [v2.43.29]
- Fixed VS.GetPlayerByUserid returning invalid player instances on disconnected players who were extended with ToExtendedPlayer
- Game string pool optimisation: disconnected player script IDs are removed on round start

2023-01-25 [v2.43.28]
- Fixed disabling player input listener ( VS.SetInputCallback(player, null, null, null) )

2022-12-21 [v2.43.27]
- Reduced stack size of VS.QuaternionSquad

2022-10-02 [v2.43.26]
- Fixed Portal 2 CExtendedPlayer save/restore crash

2022-09-16 [v2.43.25]
- Increased precision of VS.MatrixBuildRotation to 0.000061 degrees from 0.0573 degrees
- Fixed loading an old version of the library over a new one invalidating ::ToExtendedPlayer. Fix is applied post-2.43.5. Retroactive manual fix is to use VS.ToExtendedPlayer

2022-08-26 [v2.43.24]
- Added VS.ClipRayToBox
- Added VS.ClipRayToOBB
- Fixed hull and non-swept trace in VS.IsRayIntersectingOBB
- Fixed trace parameter of VS.IntersectRayWithRay
- Fixed VS.CalcSqrDistanceToAABB
- Fixed VS.CalcClosestPointOnAABB return value

2022-08-07 [v2.43.23]
- Added VS.IsOBBIntersectingOBB
- Added VS.ComputeSeparatingPlane

2022-07-12 [v2.43.22]
- Fixed VS.ConcatTransforms incorrect translation calculation
- Deprecated CExtendedPlayer::SetInputCallback(), moved to VS.SetInputCallback(). No functional changes.
	Update existing code by changing `player.SetInputCallback( ... )` to `VS.SetInputCallback( player, ... )`
- Removed VS.FindEntityNearestFacing, VS.FindEntityClassNearestFacing, VS.FindEntityClassNearestFacingNearest
- Removed VS.Hermite_SplineBasis
- Removed VS.Log.enabled
- Removed VS.Log.Run data parameter, it now only takes callback and environment parameters
- Removed VS.Log.Run call environment fallback for user callback to allow already bound functions

2022-06-25 [v2.43.21]
- Added VS.WorldToScreen
- Changed VS.WorldToScreenMatrix and VS.ComputeCameraVariables to be consistent with engine definitions. Coordinate range is changed from [0,1] to [-1,1].

2022-06-23 [v2.43.20]
- Added VS.WorldToScreenMatrix
- Removed VS.ScreenToWorldMatrix
	To get the ScreenToWorld matrix, invert the WorldToScreen matrix using VS.MatrixInverseGeneral
	Note that VS.ScreenToWorld still requires ScreenToWorld matrix input.

2022-06-09 [v2.43.19]
- Fixed vertical VS.TraceLine.GetNormal calculations
- Fixed CExtendedPlayer::SetInputCallback callback unregistration

2022-06-06 [v2.43.18]
- Fixed in==out condition in VS.MatrixInverseTR
- Removed VS.ComputeViewMatrix
- Removed trace_t metamethods

2022-06-03 [v2.43.17]
- Added CExtendedPlayer::GetFOV

2022-05-30 [v2.43.16]
- Added support for synchronous game event callbacks. See the documentation on VS.ListenToGameEvent for details.
- Added (debug) VS.Events.DumpListeners
- Removed VS.FixupEventListener
- Removed deprecated functions VS.ValidateUseridAll, VS.ForceValidateUserid

2022-05-27 [v2.43.15]
- Fixed VS.TraceLine.GetNormal() regression in v2.43.14
- Fixed CExtendedPlayer::SetInputCallback() cleanup on disabling the listener
- Fixed VS.Catmull_Rom_Spline_Integral()
- Fixed VS.Catmull_Rom_Spline_Integral2()

2022-05-14 [v2.43.14]
- Fixed VS.InterpolateAngles not having a default output parameter
- Fixed VS.QuaternionExp, VS.QuaternionLn, VS.QuaternionSquad, VS.QuaternionAverageExponential functions returning incorrect values
- Fixed VS.Hermite_SplineBasis

2022-05-11 [v2.43.13]
- Fixed VS.DrawCapsule missing lines and rotation
- Removed deprecated functions VS.CreateMeasure, VS.SetMeasure. Instead use ToExtendedPlayer() for player eye angles.

2022-05-08 [v2.43.12]
- Fixed CExtendedPlayer::GetUserID() always returning -1 when ToExtendedPlayer() was called before player spawn
on a player who was connected to the server before map change. It will now return the
correct value after the player spawns.
- Added more error case console prints in game events library.
- Removed deprecated functions VS.Events.ValidateUseridAll, VS.Events.ForceValidateUserid
- Fixed glow.nut bug where setting a glow after disabling one in the same frame would not enable the glow.

2022-04-20 [v2.43.11]
- Fixed save/restore crash in Portal 2
- Fixed VS.VectorYawRotate() regression for in==out conditions
- Added console message print on library load
- Removed support for old vs_events library methods

2021-10-20 [v2.43.10]
- Fixed VS.MatrixCopy for cross-type matrix copying (matrix3x4_t v VMatrix and vice versa)

2021-10-19 [v2.43.9]
- Fixed a bug in VS.RotationDelta that would calculate the wrong value
- Eliminated redundant code in VS.QuaternionMult
- Optimised VS.QuaternionMA
- Optimised VS.MatrixCopy
- Optimised VS.IsRayIntersectingOBB

2021-10-17 [v2.43.8]
- Added alternative parameters on VS.AddOutput (output, target, input, parameter, delay, timesToFire)
- More minor bytecode optimisations in math lib
- Fixed vs_events.nut compilation error

2021-10-10 [v2.43.7]
vs_math:
- Changed method of internal matrix storage for improved performance and memory usage on initialisation and member access.
This especially matters for places with high amount of transforms and multiplications.
For instance: VectorTransform is 25%, MatrixMultiply is 70% faster.
However 2D member access is now significantly slower than previously.
- Bytecode optimisations mostly favouring less instruction count in exchange for larger stack size.
- Bytecode optimisations to reduce stack size in exchange for lesser readability.
- Added matrix3x4_t::InitXYZ (FLU)
- Removed VS.ComputeVolume

2021-10-03 [v2.43.6]
- Fixed VS.MatrixBuildRotationAboutAxis bug in the minified files
- Fixed in==out conditions in VS.ConcatRotations
- Optimisations in VS.ConcatRotations, VS.ConcatTransforms, VS.MatrixMultiply (decreased instr, increased stack size)
- Renamed VS.ExponentialDecay2 to VS.ExponentialDecayHalf
- Optimisation in vs_events
- CExtendedPlayer member variables are now forward compatible

src/vs_library.lua
-------------------------------------------------------------------------
--------------------- Copyright (c) samisalreadytaken -------------------
-------------------------------------------------------------------------

local VER = "0.1.4"

if not _VS then
	_VS = {}
elseif _VS[VER] then
	return _VS[VER]
end

local VS = {}
_VS[VER] = VS

-------------------------------------------------------------------------

MAX_COORD_FLOAT = 16384.0
MAX_TRACE_LENGTH = 56755.84086241697115430736
DEG2RAD = 0.01745329251994329576
RAD2DEG = 57.29577951308232087679
PI = 3.14159265358979323846
RAND_MAX = 0x7FFF

local Msg,Warning = Msg,Warning
local append,select,type,f,floor = table.insert,select,type,string.format,math.floor
local unpack,pack = unpack,function(...) return { n = select("#",...),... } end

if not table.pack then table.pack = pack end
if not table.unpack then table.unpack = unpack end

local throw = function(msg)

	local info = debug.getinfo(2,"Sl")
	Warning(msg)
	Warning(f("\t%s:%d\n",info.short_src,info.currentline))
	return false

end

if SERVER_DLL then do ---------------------------------------------------

local tQueue = {}
local nCompletedInQueue = 0
local bPlayerSpawned = false

-------------------------------------------------------------------------
-- Input:  closure, string|closure error, params...
--
-- Output: true on success
-------------------------------------------------------------------------
function VS.OnPlayerSpawn(...)

	local ply = Entities:GetLocalPlayer()

	if bPlayerSpawned or ply and
	   type(ply) == "table" and
	   IsValidEntity(ply) and
	   not ply:IsNull() and
	   ply:GetClassname() == "player" then
		return Warning("VS::OnPlayerSpawn: player is already spawned\n")
	end

	local argv = pack(...)
	local func = argv[1]
	local info = debug.getinfo(func,"S")

	if argv.n == 0 or type(func) ~= "function" then
		return throw("VS::OnPlayerSpawn: invalid parameter 1 ["..info.short_src.."]\n")
	end

	for i = 1, #tQueue do
		if tQueue[tQueue[i]].source == info.source then
			return throw("VS::OnPlayerSpawn: found event from this file in the queue, aborting ["..info.short_src.."]\n")
		end
	end

	local msg = "VS::OnPlayerSpawn: closure added to queue"
	local err_cb,err_msg,params

	if argv.n > 1 then

		local typ = type(argv[2])

		if typ == "string" then
			err_msg = argv[2]
		elseif typ == "function" then
			err_cb = argv[2]
		elseif typ ~= "nil" then
			return throw("VS::OnPlayerSpawn: parameter 2 has an invalid type '"..typ.."' ; expected 'string|function'\n")
		end

		if argv.n > 2 then

			params = {}

			for i = 3, argv.n do
				append(params,argv[i])
			end

			msg = f( "%s with %d parameter(s)", msg, #params )

		end

	end

	append(tQueue,func)
	tQueue[func] =
	{
		success = false,
		params = params,
		err_cb = err_cb,
		err_msg = err_msg,
		source = info.source,
		src = info.short_src,
		line = info.linedefined
	}

	Msg(f( "%s [%s]\n", msg, info.short_src ))

	return true

end

local function RunQueue()

	local len = #tQueue

	for i = 1, len do

		local f = tQueue[i]
		local t = tQueue[f]

		if not t.success then

			local p,r = t.params

			if p then
				r = f(unpack(p))
			else
				r = f()
			end

			if r or r == nil then

				t.success = true
				nCompletedInQueue = nCompletedInQueue+1

			end

		end

	end

	if nCompletedInQueue == len then
		return true
	else
		return false
	end

end

local function FixLevelChange()

	local bIsTools = IsInToolsMode()
	local szMapName = GetMapName()

	if szMapName == "a1_intro_world" then

		local hCmd = Entities:FindByName(nil,"command_change_level")
		if hCmd then hCmd:Kill() end

		local ent = Entities:FindByName(nil,"relay_stun_player")
		if ent then

			ent:GetOrCreatePrivateScriptScope().OnTriggerLevelChange = function()

				ent:SetContextThink("VS_LevelChange",function()

					if bIsTools then
						SendToConsole("addon_tools_map a1_intro_world_2")
					else
						SendToConsole("addon_play a1_intro_world_2")
					end

				end,1.5)

			end

			ent:RedirectOutput("OnTrigger","OnTriggerLevelChange",ent)

		end

	end

end

local function Destroy()

	throw = nil
	tQueue = nil
	RunQueue = nil
	nCompletedInQueue = nil
	FixLevelChange = nil
	Destroy = nil

end

if not _VS.__eventspawn and not Entities:GetLocalPlayer()then

	_VS.__eventspawn = ListenToGameEvent("player_connect_full", function()

		bPlayerSpawned = true

		if _VS.__dummy then
			_VS.__dummy:Kill()
			_VS.__dummy = nil
		end
		_VS.__dummy = Entities:CreateByClassname("soundent")

		FixLevelChange()

		if not RunQueue() then

			local nInitCount = 0

			_VS.__dummy:SetContextThink("", function()

				nInitCount = nInitCount+1

				if RunQueue() or nInitCount > 10 then

					if nInitCount > 10 then

						Warning(f("VS::PostPlayerSpawn: timeout, failed to execute %d function(s)\n",
						                                                   #tQueue-nCompletedInQueue))

						for i = 1, #tQueue do

							local t = tQueue[tQueue[i]]

							if not t.success then

								if t.err_msg then
									Warning(t.err_msg.."\n")
								elseif t.err_cb then
									t.err_cb()
								else
									Warning(f("\t%s:%d\n",t.src,t.line))
								end

							end

						end

					end

					Destroy()
					nInitCount = nil

					-- delete next frame to preserve other event listeners
					_VS.__dummy:SetContextThink( "R", function()
						StopListeningToGameEvent(_VS.__eventspawn)
						_VS.__eventspawn = nil
						_VS.__dummy:Kill()
						_VS.__dummy = nil
					end, 0.0 )

					return

				end

				return 1.0

			end, 1.0)

			Msg("VS::PostPlayerSpawn: running ("..tostring(_VS.__dummy:entindex())..")\n")

		else

			Destroy()
			_VS.__dummy:SetContextThink( "R", function()
				StopListeningToGameEvent(_VS.__eventspawn)
				_VS.__eventspawn = nil
				_VS.__dummy:Kill()
				_VS.__dummy = nil
			end, 0.0 )

		end

	end, nil)

end

end end -- SERVER_DLL

function VS.IsAddonEnabled(str)

	for addon in Convars:GetStr("default_enabled_addons_list"):gmatch("[^,]+") do
		if addon == str then
			return true
		end
	end
	return false

end

-------------------------------------------------------------------------
-- utilsinit.lua

if package.loaded["utils.utilsinit"] then

function Deg2Rad(deg)
	return deg * 0.01745329251994329576
end

function Rad2Deg(rad)
	return rad * 57.29577951308232087679
end

function VectorDistanceSq(v1, v2)
	local l = (v1-v2):Length()
	return l*l
end

function VectorDistance(v1, v2)
	return (v1-v2):Length()
end

function VectorLerp(t, a, b)
	local c = (b-a)*t
	return a+c
end

end

-------------------------------------------------------------------------

return VS


vs_library/src.lua
-------------------------------------------------------------------------
--------------------- Copyright (c) samisalreadytaken -------------------
--- v0.1.1 --------------------------------------------------------------

local VER = "v0.1.1"

if not _VS then
	_VS = {}
elseif _VS[VER] then
	return _VS[VER]
end

local VS = {}
_VS[VER] = VS

-------------------------------------------------------------------------

VS.MAX_COORD_FLOAT = 16384.0
VS.MAX_TRACE_LENGTH = 56755.84086241697115430736
VS.DEG2RAD = 0.01745329251994329576
VS.RAD2DEG = 57.29577951308232087679
VS.PI = 3.14159265358979323846
VS.RAND_MAX = 0x7FFF

local Msg,Warning = Msg,Warning
local append,select,type,f,floor = table.insert,select,type,string.format,math.floor
local unpack,pack = unpack,function(...) return { n = select("#",...),... } end

if not table.pack then table.pack = pack end
if not table.unpack then table.unpack = unpack end

local throw = function(msg)

	local info = debug.getinfo(2,"Sl")
	Warning(msg)
	Warning(f("\t%s:%d\n",info.short_src,info.currentline))
	return false

end

if IsServer() then do ---------------------------------------------------

local tQueue = {}
local nCompletedInQueue = 0
local bPlayerSpawned = false

-------------------------------------------------------------------------
-- Input:  closure, string|closure error, params...
--
-- Output: true on success
-------------------------------------------------------------------------
function VS.OnPlayerSpawn(...)

	local ply = Entities:GetLocalPlayer()

	if bPlayerSpawned or ply and
	   type(ply) == "table" and
	   type(ply.__self) == "userdata" and
	   not ply:IsNull() and
	   ply:GetClassname() == "player" then
		return Warning("VS::OnPlayerSpawn: player is already spawned\n")
	end

	local argv = pack(...)
	local func = argv[1]

	if argv.n == 0 or type(func) ~= "function" then
		return throw("VS::OnPlayerSpawn: invalid parameter 1\n")
	end

	local info = debug.getinfo(func,"S")

	for i = 1, #tQueue do
		if tQueue[tQueue[i]].source == info.source then
			return throw("VS::OnPlayerSpawn: found event from this file in the queue, aborting\n")
		end
	end

	local msg = "VS::OnPlayerSpawn: closure added to queue"
	local err_cb,err_msg,params

	if argv.n > 1 then

		local typ = type(argv[2])

		if typ == "string" then
			err_msg = argv[2]
		elseif typ == "function" then
			err_cb = argv[2]
		elseif typ ~= "nil" then
			return throw("VS::OnPlayerSpawn: parameter 2 has an invalid type '"..typ.."' ; expected 'string|function'\n")
		end

		if argv.n > 2 then

			params = {}

			for i = 3, argv.n do
				append(params,argv[i])
			end

			msg = msg.." with "..tostring(#params).." parameter(s)"

		end

	end

	append(tQueue,func)
	tQueue[func] =
	{
		bSuccess = false,
		params = params,
		err_cb = err_cb,
		err_msg = err_msg,
		source = info.source,
		src = info.short_src,
		line = info.linedefined
	}

	Msg(msg.."\n")

	return true

end

local function RunQueue()

	local len = #tQueue

	for i = 1, len do

		local f = tQueue[i]
		local t = tQueue[f]

		if not t.bSuccess then

			local p,r = t.params

			if p then
				r = f(unpack(p))
			else
				r = f()
			end

			if r or r == nil then

				t.bSuccess = true
				nCompletedInQueue = nCompletedInQueue+1

			end

		end

	end

	if nCompletedInQueue == len then
		return true
	else
		return false
	end

end

local function FixLevelChange()

	local bIsTools = IsInToolsMode()
	local szMapName = GetMapName()

	if szMapName == "a1_intro_world" then

		local hCmd = Entities:FindByName(nil,"command_change_level")
		if hCmd then hCmd:Kill() end

		local ent = Entities:FindByName(nil,"relay_stun_player")
		if ent then

			ent:GetOrCreatePrivateScriptScope().OnTriggerLevelChange = function()

				ent:SetContextThink("VS_LevelChange",function()

					if bIsTools then
						SendToConsole("addon_tools_map a1_intro_world_2")
					else
						SendToConsole("addon_play a1_intro_world_2")
					end

				end,1.5)

			end

			ent:RedirectOutput("OnTrigger","OnTriggerLevelChange",ent)

		end

	end

end

local function Destroy()

	throw = nil
	tQueue = nil
	RunQueue = nil
	nCompletedInQueue = nil
	FixLevelChange = nil
	Destroy = nil

end

if not _VS.__iEventSpawn and not Entities:GetLocalPlayer()then

	_VS.__iEventSpawn = ListenToGameEvent("player_connect_full", function()

		bPlayerSpawned = true

		if _VS.__hSpawnInit then
			_VS.__hSpawnInit:Kill()
			_VS.__hSpawnInit = nil
		end

		FixLevelChange()

		if not RunQueue() then

			local nInitCount = 0

			_VS.__hSpawnInit = Entities:CreateByClassname("soundent")

			_VS.__hSpawnInit:SetContextThink("",function()

				nInitCount = nInitCount+1

				if RunQueue() or nInitCount > 10 then

					if nInitCount > 10 then

						Warning(f("VS::PostPlayerSpawn: timeout, failed to execute %d function(s)\n",
						                                                   #tQueue-nCompletedInQueue))

						for i = 1, #tQueue do

							local t = tQueue[tQueue[i]]

							if not t.bSuccess then

								if t.err_msg then
									Warning(t.err_msg.."\n")
								elseif t.err_cb then
									t.err_cb()
								else
									Warning(f("\t%s:%d\n",t.src,t.line))
								end

							end

						end

					end

					Destroy()
					nInitCount = nil
					_VS.__hSpawnInit:Kill()
					_VS.__hSpawnInit = nil
					StopListeningToGameEvent(_VS.__iEventSpawn)
					_VS.__iEventSpawn = nil

					return

				end

				return 1.0

			end,1.0)

			Msg("VS::PostPlayerSpawn: running ("..tostring(_VS.__hSpawnInit:entindex())..")\n")

		else

			Destroy()
			StopListeningToGameEvent(_VS.__iEventSpawn)
			_VS.__iEventSpawn = nil

		end

	end, nil)

end

end end -- IsServer

function VS.IsAddonEnabled(str)

	for addon in Convars:GetStr("default_enabled_addons_list"):gmatch("[^,]+") do
		if addon == str then
			return true
		end
	end
	return false

end

-------------------------------------------------------------------------
-- math.nut

function VS.IsInteger(f) return floor(f) == f end

function VS.IsLookingAt( vSrc, vTarget, vDir, cosTolerance )

	return (vTarget-vSrc):Normalized():Dot(vDir) >= cosTolerance

end

function VS.PointOnLineNearestPoint( vStartPos, vEndPos, vPoint )

	local v1 = vEndPos - vStartPos
	local v1l = v1:Length()
	local dist = v1:Dot(vPoint-vStartPos) / (v1l*v1l)

	if dist < 0.0 then
		return vStartPos
	elseif dist > 1.0 then
		return vEndPos
	else
		return vStartPos + v1 * dist
	end

end

function VS.Approach( tg, val, spd )

	local dt = tg - val

	if dt > spd then
		val = val + spd
	elseif dt < -spd then
		val = val - spd
	else
		val = tg
	end

	return val

end

function VS.ApproachAngle( tg, val, spd )

	tg = tg % 360.0

	if tg > 180.0 then
		tg = tg - 360.0
	elseif tg < -180.0 then
		tg = tg + 360.0
	end

	val = val % 360.0

	if val > 180.0 then
		val = val - 360.0
	elseif val < -180.0 then
		val = val + 360.0
	end

	local dt = tg - val

	dt = dt % 360.0

	if dt > 180.0 then
		dt = dt - 360.0
	elseif dt < -180.0 then
		dt = dt + 360.0
	end

	if spd < 0 then
		spd = -spd
	end

	if dt > spd then
		val = val + spd
	elseif dt < -spd then
		val = val - spd
	else
		val = tg
	end

	return val

end

function VS.AngleDiff( dst, src )

	local ang = dst - src

	ang = ang % 360.0

	if ang > 180.0 then
		ang = ang - 360.0
	elseif ang < -180.0 then
		ang = ang + 360.0
	end

	return ang

end

function VS.AngleNormalize( ang )

	ang = ang % 360.0

	if ang > 180.0 then
		ang = ang - 360.0
	elseif ang < -180.0 then
		ang = ang + 360.0
	end

	return ang

end

function VS.QAngleNormalize( ang )

	ang.x = ang.x % 360.0
	ang.y = ang.y % 360.0
	ang.z = ang.z % 360.0
	if ang.x > 180.0 then
		ang.x = ang.x - 360.0
	elseif ang.x < -180.0 then
		ang.x = ang.x + 360.0
	end
	if ang.y > 180.0 then
		ang.y = ang.y - 360.0
	elseif ang.y < -180.0 then
		ang.y = ang.y + 360.0
	end
	if ang.z > 180.0 then
		ang.z = ang.z - 360.0
	elseif ang.z < -180.0 then
		ang.z = ang.z + 360.0
	end

end

function VS.SnapDirectionToAxis( vDir, eps )

	local proj = 1.0 - eps

	local x if vDir.x < 0 then x = -vDir.x else x = vDir.x end

	if x > proj then

		if vDir.x < 0.0 then
			vDir.x = -1.0
		else
			vDir.x = 1.0
		end
		vDir.y = 0.0
		vDir.z = 0.0

		return vDir

	end

	local y if vDir.y < 0 then y = -vDir.y else y = vDir.y end

	if y > proj then

		if vDir.y < 0.0 then
			vDir.y = -1.0
		else
			vDir.y = 1.0
		end
		vDir.z = 0.0
		vDir.x = 0.0

		return vDir

	end

	local z if vDir.z < 0 then z = -vDir.z else z = vDir.z end

	if z > proj then

		if vDir.z < 0.0 then
			vDir.z = -1.0
		else
			vDir.z = 1.0
		end
		vDir.x = 0.0
		vDir.y = 0.0

		return vDir

	end

end

function VS.VectorsAreEqual( a, b, tolerance )

	if not tolerance then tolerance = 0 end

	local x = a.x - b.x
	if x < 0 then x = -x end

	local y = a.y - b.y
	if y < 0 then y = -y end

	local z = a.z - b.z
	if z < 0 then z = -z end

	return ( x <= tolerance and
	         y <= tolerance and
	         z <= tolerance )
end

-------------------------------------------------------------------------
-- utilsinit.lua

if package.loaded["utils.utilsinit"] then

function Deg2Rad(deg)
	return deg * 0.01745329251994329576
end

function Rad2Deg(rad)
	return rad * 57.29577951308232087679
end

function VectorDistanceSq(v1, v2)
	local l = (v1-v2):Length()
	return l*l
end

function VectorDistance(v1, v2)
	return (v1-v2):Length()
end

function VectorLerp(t, a, b)
	local c = (b-a)*t
	return a+c
end

end

-------------------------------------------------------------------------

return VS
